### ISSUE TITLE
SEACAS:  Undefined symbols for architecture x86_64

### ISSUE THREAD

**Issue by jmgate** [trilinos/Trilinos#5644]
## Question
@trilinos/seacas 

I'm attempting to build Trilinos on Mac OS-X for the first time, and I'm running into a build error I've never seen before:
```
[ 42%] Linking CXX executable SEACASNemesis_ne_ctest_wrap.exe
Undefined symbols for architecture x86_64:
  "_ncmpi_inq_libvers", referenced from:
      _ex_print_config in libexodus.a(ex_utils.c.o)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
```

What am I doing wrong?  Thanks in advance for any help.

### Environment
Mac OS-X 10.14.6 with
- boost-1.70.0
- cmake-3.15.0
- gcc-9.1.0
- hdf5-1.10.5
- lapack-3.8.0
- netcdf-c-4.7.0
- openmpi-4.0.1
- pnetcdf-1.11.2
- zlib-1.2.11

### Scripts / Logs
* [configure script](https://github.com/trilinos/Trilinos/files/3462604/configure.txt)
* [configure log](https://github.com/trilinos/Trilinos/files/3462607/config190802121759.log)
* [build log](https://github.com/trilinos/Trilinos/files/3462611/make190802131005.log)

> **Note:**  If the configure script looks a little odd, it's because I'm building Trilinos as part of the Charon codebase.  This same method of configuring and building works just fine on my RHEL7 machine.

**Comment by jmgate** [comment]
@mhoemmen, do you by chance know who I should ping about this?

**Comment by rppawlo** [comment]
@gsjaardema

**Comment by mhoemmen** [comment]
@jmgate sorry, was in a phone meeting.  Roger has the right idea :-)

**Comment by mayrmt** [comment]
I'm having  a similar linker error on Mac 10.14.6 with Clang 10.0.1 using [do-configure-trilinos_minimal.txt](https://github.com/trilinos/Trilinos/files/3479171/do-configure-trilinos_minimal.txt) and OpenMPI 3.1.4 installed via [spack](spack.io).

Here's the terminal output:
```
[325/496] Linking CXX shared library packages/xpetra/src/libxpetra.12.17.dylib
FAILED: packages/xpetra/src/libxpetra.12.17.dylib 
: && /Users/mayr/utils/spack/opt/spack/darwin-mojave-x86_64/clang-10.0.1-apple/openmpi-3.1.4-qz6zsrmc4l4tdlthlxppwr34bmdxlswa/bin/mpic++ -fdiagnostics-color=always -w -std=c++11 -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 12.0.0 -current_version 12.17.0 -o packages/xpetra/src/libxpetra.12.17.dylib -install_name @rpath/libxpetra.12.dylib packages/xpetra/src/CMakeFiles/xpetra.dir/Platform/Xpetra_DefaultPlatform.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/Utils/Xpetra_EpetraUtils.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/Utils/Xpetra_Utils.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/CrsGraph/Xpetra_EpetraCrsGraph.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/CrsMatrix/Xpetra_EpetraCrsMatrix.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/Export/Xpetra_EpetraExport.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/Import/Xpetra_EpetraImport.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/Map/Xpetra_EpetraMap.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/MultiVector/Xpetra_EpetraIntMultiVector.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/MultiVector/Xpetra_EpetraMultiVector.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/MultiVector/Xpetra_EpetraMultiVectorFactory.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/Vector/Xpetra_EpetraIntVector.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/Vector/Xpetra_EpetraVector.cpp.o packages/xpetra/src/CMakeFiles/xpetra.dir/BlockedCrsMatrix/Xpetra_BlockReorderManager.cpp.o  -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/epetraext/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/epetra/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/kokkoscomm/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/kokkoscompat/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/remainder/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/numerics/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/comm/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/parameterlist/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/parser/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/teuchos/core/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/kokkos/containers/src -Wl,-rpath,/Users/mayr/research/codes/mayrmt_trilinos/build_minimal/packages/kokkos/core/src packages/epetraext/src/libepetraext.12.17.dylib packages/epetra/src/libepetra.12.17.dylib packages/teuchos/kokkoscomm/src/libteuchoskokkoscomm.12.17.dylib packages/teuchos/kokkoscompat/src/libteuchoskokkoscompat.12.17.dylib packages/teuchos/remainder/src/libteuchosremainder.12.17.dylib packages/teuchos/numerics/src/libteuchosnumerics.12.17.dylib packages/teuchos/comm/src/libteuchoscomm.12.17.dylib packages/teuchos/parameterlist/src/libteuchosparameterlist.12.17.dylib packages/teuchos/parser/src/libteuchosparser.12.17.dylib packages/teuchos/core/src/libteuchoscore.12.17.dylib packages/kokkos/containers/src/libkokkoscontainers.12.17.dylib packages/kokkos/core/src/libkokkoscore.12.17.dylib /usr/lib/liblapack.dylib /usr/lib/libblas.dylib /usr/lib/libdl.dylib && :
Undefined symbols for architecture x86_64:
  "Xpetra::BlockedMultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::BlockedMultiVector(Teuchos::RCP<Xpetra::BlockedMap<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, unsigned long, bool)", referenced from:
      Xpetra::MultiVectorFactory<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Build(Teuchos::RCP<Xpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, unsigned long, bool) in Xpetra_EpetraMultiVectorFactory.cpp.o
  "Xpetra::BlockedMultiVector<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::BlockedMultiVector(Teuchos::RCP<Xpetra::BlockedMap<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, unsigned long, bool)", referenced from:
      Xpetra::MultiVectorFactory<int, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Build(Teuchos::RCP<Xpetra::Map<int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, unsigned long, bool) in Xpetra_EpetraMultiVectorFactory.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by gsjaardema** [comment]
It looks like the `SEACASNemesis_ne_ctest_wrap.exe` is not picking up the `PnetCDF`dependency on the Exodus library.  I will take a look and see if I can figure out why.  Note that it works fine on my Mac system in parallel and serial...

Is this a parallel (MPI-enabled) build?

**Comment by jmgate** [comment]
> Is this a parallel (MPI-enabled) build?

Yes.  The configure script and logs are linked in the description.

**Comment by gsjaardema** [comment]
If you do 'grep TPL_Netcdf CMakeCache.txt` do you get something like the following:
```
TPL_Netcdf_Enables_Netcdf4:BOOL=True
TPL_Netcdf_Enables_PNetcdf:BOOL=True
TPL_Netcdf_INCLUDE_DIRS:PATH=/Users/gdsjaar/src/seacas-parallel/include;/Users/gdsjaar/src/seacas-parallel/include;/Users/gdsjaar/src/seacas-parallel/include
TPL_Netcdf_LIBRARIES:PATH=/Users/gdsjaar/src/seacas-parallel/lib/libnetcdf.dylib;/Users/gdsjaar/src/seacas-parallel/lib/libhdf5_hl.dylib;/Users/gdsjaar/src/seacas-parallel/lib/libhdf5.dylib;/usr/lib/libz.dylib;/usr/lib/libdl.dylib;/usr/lib/libm.dylib;/Users/gdsjaar/src/seacas-parallel/lib/libpnetcdf.dylib
TPL_Netcdf_PARALLEL:BOOL=True
TPL_Netcdf_LIBRARY_DIRS:INTERNAL=
TPL_Netcdf_NOT_FOUND:INTERNAL=FALSE
```
I think the important one is the `TPL_Netcdf_Enables_PNetcdf` setting.

**Comment by jmgate** [comment]
I get
```
// any of the TPL_Netcdf_[INCLUDE_DIRS,LIBRARY_NAMES,LIBRARY_DIRS]
// of libraries used for linking.  That is specified by TPL_Netcdf_LIBRARIES!
TPL_Netcdf_INCLUDE_DIRS:PATH=/Users/jmgate/toolchain/install/netcdf-c-4.7.0/include
TPL_Netcdf_LIBRARIES:FILEPATH=/Users/jmgate/toolchain/install/netcdf-c-4.7.0/lib/libnetcdf.dylib
//ADVANCED property for variable: TPL_Netcdf_INCLUDE_DIRS
TPL_Netcdf_INCLUDE_DIRS-ADVANCED:INTERNAL=1
//ADVANCED property for variable: TPL_Netcdf_LIBRARIES
TPL_Netcdf_LIBRARIES-ADVANCED:INTERNAL=1
TPL_Netcdf_LIBRARY_DIRS:INTERNAL=
TPL_Netcdf_NOT_FOUND:INTERNAL=FALSE
```

Looks like a few things are mismatched, but I'm not sure what to do about them.

**Comment by gsjaardema** [comment]
Yes, there are definitely some issues here.  My guess is that the `netcdf_meta.h` file probably says that your NetCDF library depends on PnetCDF and possibly HDF5, but those libraries are not being set as dependencies of NetCDF.  

The configure.log seems to be bypassing the NetCDF code that sets the options. For example, my configure outputs this:
```
Processing enabled TPL: Netcdf (enabled explicitly, disable with -DTPL_ENABLE_Netcdf=OFF)
-- Using FIND_PACKAGE(Netcdf ...) ...
-- NetCDF include file /Users/gdsjaar/src/seacas-parallel/include/netcdf.h will be searched for define values
-- 	NetCDF_ROOT is /Users/gdsjaar/src/seacas-parallel
-- Found NetCDF configuration script: /Users/gdsjaar/src/seacas-parallel/bin/nc-config
-- /Users/gdsjaar/src/seacas-parallel/bin/nc-config --has-hdf5 returned 'yes'
-- /Users/gdsjaar/src/seacas-parallel/bin/nc-config --has-pnetcdf returned 'yes'
-- NetCDF requires HDF5
-- NetCDF depends on HDF5
-- Updating NetCDF_LIBRARIES and NetCDF_INCLUDE_DIRS
```
I'm not sure what is causing the difference.  Maybe @bartlettroscoe has some insight.

**Comment by jmgate** [comment]
Here's what I've got:
```
Processing enabled TPL: Netcdf (enabled explicitly, disable with -DTPL_ENABLE_Netcdf=OFF)
-- Netcdf_LIBRARY_NAMES='netcdf'
-- Searching for libs in Netcdf_LIBRARY_DIRS='/Users/jmgate/toolchain/install/netcdf-c-4.7.0/lib'
-- Searching for a lib in the set "netcdf":
--   Searching for lib 'netcdf' ...
--     Found lib '/Users/jmgate/toolchain/install/netcdf-c-4.7.0/lib/libnetcdf.dylib'
-- TPL_Netcdf_LIBRARIES='/Users/jmgate/toolchain/install/netcdf-c-4.7.0/lib/libnetcdf.dylib'
-- Searching for headers in Netcdf_INCLUDE_DIRS='/Users/jmgate/toolchain/install/netcdf-c-4.7.0/include'
-- Searching for a header file in the set "netcdf.h":
--   Searching for header 'netcdf.h' ...
--     Found header '/Users/jmgate/toolchain/install/netcdf-c-4.7.0/include/netcdf.h'
-- Found TPL 'Netcdf' include dirs '/Users/jmgate/toolchain/install/netcdf-c-4.7.0/include'
-- TPL_Netcdf_INCLUDE_DIRS='/Users/jmgate/toolchain/install/netcdf-c-4.7.0/include'
```

**Comment by gsjaardema** [comment]
Can you run `/Users/jmgate/toolchain/install/netcdf-c-4.7.0/bin/nc-config --all` and send the output.

**Comment by jmgate** [comment]
```
This netCDF 4.7.0 has been built with the following features: 

  --cc            -> /Users/jmgate/toolchain/install/openmpi-4.0.1/bin/mpicc
  --cflags        -> -I/Users/jmgate/toolchain/install/netcdf-c-4.7.0/include 
  --libs          -> -L/Users/jmgate/toolchain/install/netcdf-c-4.7.0/lib -lnetcdf
  --static        -> -lpnetcdf -lhdf5_hl -lhdf5 -lm -lz -lcurl 

  --has-c++       -> no
  --cxx           -> 

  --has-c++4      -> no
  --cxx4          -> 

  --has-fortran   -> no
  --has-dap       -> yes
  --has-dap2      -> yes
  --has-dap4      -> yes
  --has-nc2       -> yes
  --has-nc4       -> yes
  --has-hdf5      -> yes
  --has-hdf4      -> no
  --has-logging   -> no
  --has-pnetcdf   -> yes
  --has-szlib     -> no
  --has-cdf5      -> yes
  --has-parallel4 -> yes
  --has-parallel  -> yes

  --prefix        -> /Users/jmgate/toolchain/install/netcdf-c-4.7.0
  --includedir    -> /Users/jmgate/toolchain/install/netcdf-c-4.7.0/include
  --libdir        -> /Users/jmgate/toolchain/install/netcdf-c-4.7.0/lib
  --version       -> netCDF 4.7.0
```

You're thinking something went wrong in the NetCDF build/install?  Thanks for helping with this.

**Comment by gsjaardema** [comment]
No, the build looks good.  For some reason CMake/TriBits is not picking up the PnetCDF and HDF5 dependencies of NetCDF and therefore not adding those libraries to the link lines and that is why you are getting the undefined symbols.  

My guess is that if you do:
* cd packages/seacas/libraries/nemesis
* make VERBOSE=1

Then the link line for `SEACASNemesis_ne_ctest_wrap.exe` will only show `libexodus.a` and `libnetcdf.a` and not the `libpnetcdf.a` and `libhdf5_hl.a` and `libhdf5.a`  (or `.dylib` if shared libs)

**Comment by jmgate** [comment]
```
/Users/jmgate/toolchain/install/openmpi-4.0.1/bin/mpicxx \
  -g \
  -g \
  -Wall \
  -Wextra \
  -Wpedantic \
  -std=c++11 \
  -g \
  -O0 \
  -Wl,-search_paths_first \
  -Wl,-headerpad_max_install_names \
  CMakeFiles/SEACASNemesis_ne_ctest_wrap.dir/ne_ctest_wrap.c.o \
  -o SEACASNemesis_ne_ctest_wrap.exe \
  libnemesis.a \
  ../exodus/libexodus.a \
  /Users/jmgate/toolchain/install/netcdf-c-4.7.0/lib/libnetcdf.dylib \
  /Users/jmgate/toolchain/install/hdf5-1.10.5/lib/libhdf5.dylib \
  /usr/lib/libz.dylib \
  /Users/jmgate/toolchain/install/hdf5-1.10.5/lib/libhdf5_hl.dylib 
```

**Comment by gsjaardema** [comment]
OK, it is getting the hdf5 libs, but not the PnetCDF libs...

Is it possible to reconfigure and change the lines
```
-D Netcdf_INCLUDE_DIRS:FILEPATH="${NETCDF_BASE_DIR}/include" \
-D Netcdf_LIBRARY_DIRS:FILEPATH="${NETCDF_BASE_DIR}/lib" \
```
to be:
```
-D NetCDF_DIR:PATH="${NETCDF_BASE_DIR}" \
```
You may have to rm CMakeCache.txt.  Then see if the configure information for NetCDF is different...

**Comment by jmgate** [comment]
I think I'm past it now.  I had to swap out
```
-D Netcdf_INCLUDE_DIRS:FILEPATH="${NETCDF_BASE_DIR}/include" \
-D Netcdf_LIBRARY_DIRS:FILEPATH="${NETCDF_BASE_DIR}/lib" \
```
with
```
-D NetCDF_ROOT:PATH="${NETCDF_BASE_DIR}" \
```
and then also had to add a
```
-D PNetCDF_ROOT:PATH="${PNETCDF_BASE_DIR}" \
```
How I've managed for years without `PNetCDF_ROOT` on Linux is beyond me.  Thanks so much, @gsjaardema.  No way I could've figured this out without you.

**Comment by bartlettroscoe** [comment]
As shown [here](http://compsim-dashboard.sandia.gov/cdash/index.php?project=SPARC&date=2019-08-08&filtercount=1&showfilters=1&field1=buildname&compare1=66&value1=-trildev), looks like SPARC is seeing a bunch of undefined symbols in the builds:

* sparc-alltpls_cee-cpu_clang-5.0.1_openmpi-1.10.2_static_opt-trildev
* sparc-alltpls_cee-cpu_intel-17.0.1_intelmpi-5.1.2_static_opt-trildev
* sparc-alltpls_cee-cpu_intel-18.0.2_mpich2-3.2_static_opt-trildev
* sparc-alltpls_waterman-v100_cuda-9.2.88-gcc-7.2.0_openmpi-2.1.2_shared_opt-trildev

but not the build:

* sparc-alltpls_cee-cpu_gcc-7.2.0_openmpi-1.10.2_shared_opt-trildev

@jmgate, can you post a PR that fixes the ATDMDevEnvSettings.cmake file for this

**Comment by jmgate** [comment]
@bartlettroscoe, I wasn't using the ATDM config scripts when trying to build on Mac here.

**Comment by jmgate** [comment]
@bartlettroscoe, I'd suggest opening a separate issue for the problems you're seeing.

**Comment by bartlettroscoe** [comment]
@jmgate said:

> I'd suggest opening a separate issue for the problems you're seeing.

Okay, will do.
